alea/include/alps/alea/core.hpp

```diff
@@ -37,39 +37,95 @@ template <typename T>
 struct traits;
 
 /**
- * Data sink as a thin wrapper around a continuous array.
+ * Data view as a thin wrapper around a continuous array.
  *
- * Basically collects a pointer continuous array (vector, Eigen array, C
- * array etc.) together with its size.  This is sufficiently generic for the
- * virtual interface of `estimator`.
+ * Basically collects a pointer continuous array together with its size.  Note
+ * that the view does not own the `data` pointer.
  */
 template <typename T>
-class sink
+class view
 {
 public:
     typedef T value_type;
 
 public:
     /** Construct view on nothing */
-    sink() : data_(nullptr), size_(0) { }
+    view() : data_(nullptr), size_(0) { }
 
     /** Construct view on data area with size */
-    sink(T *data, size_t size) : data_(data), size_(size) { }
+    view(T *data, size_t size)
+        : data_(data), size_(size)
+    { }
 
     /** Get pointer to zeroth element */
     T *data() { return data_; }
 
     /** Get pointer to zeroth element */
     const T *data() const { return data_; }
 
-    /** Get size of data space */
+    /** Return size */
     size_t size() const { return size_; }
 
 private:
     T *data_;
     size_t size_;
 };
 
+/**
+ * Data view as a thin wrapper around a continuous multi-dimensional array.
+ *
+ * Basically collects a pointer continuous array in ROW-MAJOR format (vector,
+ * transposed Eigen array, C array etc.) together with its shape.  Note that
+ * the view neither owns the `data` pointer nor the `shape` pointer.  Thus it
+ * amends `view<T>` with shape information.
+ */
+template <typename T>
+class ndview
+    : public view<T>
+{
+public:
+    typedef T value_type;
+
+public:
+    /** Construct view on nothing */
+    ndview() : view<T>(), shape_(nullptr), ndim_(0) { }
+
+    /** Construct view on data area with shape */
+    ndview(T *data, const size_t *shape, size_t ndim)
+        : view<T>(data, compute_size(shape, ndim))
+        , shape_(shape)
+        , ndim_(ndim)
+    { }
+
+    /** Construct view on data area with shape and size hint */
+    ndview(T *data, size_t size, const size_t *shape, size_t ndim)
+        : view<T>(data, size)
+        , shape_(shape)
+        , ndim_(ndim)
+    {
+        assert(size == compute_size(shape, ndim));
+    }
+
+    /** Get shape of data space */
+    const size_t *shape() const { return shape_; }
```

Comment:
and again, why not array or vector? it at least will guarantee, that in `compute_size` you will not go out of shape allocation